### PR TITLE
Added CardFlip Animation

### LINKED ISSUE #917: Add more animations to the app
Animations in Flutter are easy and fantastic!

Check out [the official guide](https://flutter.io/docs/development/ui/animations) and [tutorial](https://flutter.io/docs/development/ui/animations/tutorial), [the boring show](https://www.youtube.com/watch?v=dNSteCm-cEY) and other [videos](https://www.youtube.com/watch?v=5urRyqOwTuo) and [articles](https://medium.com/@tonyowen/flutter-multiple-flare-animations-bf5d95d150f3) for inspiration.

We could start with animating "next card" in the learning session - like scrolling a page in a paper book, or similar. Build a list of ideas here:

- [ ] Next Card animation

### PR DESCRIPTION
**Pull request by jddeep**
Added CardFlip Animation.

ezgif com-video-to-gif

Directed to Fix #917 .

### CODE REVIEW

**Comment by jddeep** [comment]
@ksheremet Please review.

**Comment by jddeep** [comment]
@ksheremet Why has the CI build failed here?

**Comment by ksheremet** [comment]
@jddeep Please have a look at why CI failed and fix problems

**Comment by jddeep** [comment]
@ksheremet there is no build error on my local machine. Can you please help me about what is the ci error?

**Comment by jddeep** [comment]
@ksheremet this time a different error is there in the ci build. ` 
FastlaneCore::Interface::FastlaneShellError: [!] Exit status of command 'cd .. && flutter packages get' was 1 instead of 0. ` What do you suggest can be done?

**Comment by dotdoom** [comment]
@jddeep I think it was a temporary network issue on the build server. I've restarted the build.

**Comment by jddeep** [comment]
@dotdoom Thanks a lot. It really was frustrating.

**Comment by jddeep** [comment]
@ksheremet please review the PR.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 1]
```diff
@@ -17,42 +17,80 @@ class CardDisplayWidget extends StatelessWidget {
...
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: _startAnimation,
+                  child: IndexedStack(
+                    children: <Widget>[
+                      Card(
+                        color: Colors.greenAccent,
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: Color of the card is hardcoded. It is not always green.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 2]
```diff
@@ -17,42 +17,80 @@ class CardDisplayWidget extends StatelessWidget {
...
+                        child: Container(
+                            height: 300.0,
+                            width: 300.0,
+                            child: Center(
+                              child: Text(widget.front,
+                                  style: AppStyles.primaryText),
+                            )),
+                      ),
+                      Card(
+                        color: Colors.greenAccent,
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: Color of the card is hardcoded. It is not always green.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 3]
```diff
@@ -17,42 +17,80 @@ class CardDisplayWidget extends StatelessWidget {
...
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: _startAnimation,
+                  child: IndexedStack(
+                    children: <Widget>[
+                      Card(
+                        color: Colors.greenAccent,
+                        child: Container(
+                            height: 300.0,
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: Height and width are hardcoded. What if a device is too small. If a screen is big, a lot of space are unused. How about to count it dynamically depending on the size of the screen. Please have a look at MediaQuery for this.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 4]
```diff
@@ -17,42 +17,80 @@ class CardDisplayWidget extends StatelessWidget {
...
+          builder: (context, child) => Transform(
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: _startAnimation,
+                  child: IndexedStack(
+                    children: <Widget>[
+                      Card(
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: Code for creating Card widget is repeated. How about to have a method for this.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 5]
```diff
@@ -17,42 +17,80 @@ class CardDisplayWidget extends StatelessWidget {
...
+        AnimationController(vsync: this, duration: const Duration(seconds: 1));
+    _animation = TweenSequence([
+      TweenSequenceItem(tween: Tween(begin: 0.0, end: -pi / 2), weight: 0.5),
+      TweenSequenceItem(tween: Tween(begin: pi / 2, end: 0.0), weight: 0.5)
+    ]).animate(_controller);
   }
 
-  Widget _sideText(String text, BuildContext context) {
-    if (isMarkdown) {
-      return buildNonScrollingMarkdown(text, context);
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: As I see code that supports markdown feature was deleted. Is the markdown feature supported?
Please have a look on all supported features in [Readme](https://github.com/dasfoo/delern)

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 6]
```diff
@@ -1,8 +1,8 @@
+import 'dart:math';
 import 'package:delern_flutter/flutter/styles.dart';
-import 'package:delern_flutter/views/helpers/non_scrolling_markdown.dart';
 import 'package:flutter/material.dart';
 
-class CardDisplayWidget extends StatelessWidget {
+class CardDisplayWidget extends StatefulWidget {
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: As I remember that widget is used with the learning of cards (when a user taps on deck). Please make sure that learning works correctly.

Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: Yes learning works good.

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 5 above)
Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: I will look into that.

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 4 above)
Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: sure

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 1 above)
Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: So what do you suggest to use for coloring the cards @ksheremet ?

Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: Leave it i understood. i will make a change. Thanks

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 3 above)
Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: Sure

**Comment by jddeep** [comment]
@ksheremet All changes have been seen into. The other features such as Markdown have also been checked into as "Card with Container" is used here there will be no problem. Earlier there was just a nonScrollview had been created for markdown with the sideText. Here the problem has been resolved using a Container with MediaQuery heights and widths.

**Comment by jddeep** [comment]
@ksheremet Thanks.

**Comment by jddeep** [comment]
@ksheremet All changes have been seen into. The other features such as Markdown have also been checked into as "Card with Container" is used here there will be no problem. Earlier there was just a nonScrollview had been created for markdown with the sideText. Here the problem has been resolved using a Container with MediaQuery heights and widths.

**Comment by jddeep** [comment]
@ksheremet All changes are seen into. Please review the PR.

**Comment by jddeep** [comment]
@ksheremet Please review the PR moving ahead.

**Comment by ksheremet** [comment]
@jddeep Markdown functionality is broken and what if a text on a card is big?
screenshot_1551970327

**Comment by jddeep** [comment]
@ksheremet Thanks for pointing out these. I will be looking into it. Thanks.

**Comment by ksheremet** [comment]
@jddeep Flip button is useless now. Please delete 'flip button' and when card was turned to back side, show this 'know' and 'don't know' buttons. 
ezgif com-video-to-gif

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 7]
```diff
@@ -17,42 +17,72 @@ class CardDisplayWidget extends StatelessWidget {
...
+                    ],
+                    index: _controller.value < 0.5 ? 0 : 1,
+                    alignment: Alignment.center,
+                  ),
+                ),
+              ),
+        ),
+      );
+
+  Widget delernCard(Color backcolor, String text) => Padding(
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: @jddeep Method must be private. 
Please use meaningful names for methods. For example: _buildCard.
Please use camelCase and meaningul names for parameters: backcolor -> backgroundColor

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 8]
```diff
@@ -17,42 +17,72 @@ class CardDisplayWidget extends StatelessWidget {
...
+      padding: const EdgeInsets.only(top: 30, bottom: 30.0),
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: @jddeep 
`EdgeInsets.only(top: 30, bottom: 30.0)`

Inconsistency. Please use either integer or fractional

**Comment by jddeep** [comment]
ezgif com-video-to-gif
@ksheremet The Markdown feature and all other formatting changes have been seen into.

**Comment by jddeep** [comment]
@ksheremet Can you please take a look at my comment at card_display_widget where i want to use a Widget function from another class. Please help me regarding that @ksheremet @dotdoom

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 9]
```diff
@@ -1,58 +1,104 @@
...
+          builder: (context, child) => Transform(
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: () { 
+                    _startAnimation();
+                    if(backshown){
+                      // Want to call buildButtons from CardsLearning
```
Comment (dotdoom) [inline flutter/lib/views/helpers/card_display_widget.dart]: The Flutter architecture does not allow to directly update parent widget from a child.

One way to achieve this is to pass a callback into `CardDisplayWidget` (a callback is a function, like `onTap` that's passed to `GestureDetector`). The callback function body will be inside `CardsLearningState` where it can simply call `setState()`, triggering `build()` and updating button state.

Another way is to put all this code into `CardsLearningState` and wrap the whole `CardDisplayWidget` in a `GestureDetector`, then pass animation to `CardDisplayWidget` when it's flipped, similar to [this](https://github.com/dasfoo/delern/blob/2785d6f715801e6509b4a436ad3b718fc9ff19d9/flutter/lib/views/decks_list/decks_list.dart#L135-L139).

There might be more ways of doing this.

I defer to @ksheremet to make a decision, as it's her area of expertise and she's already taking care of this PR.

**Comment by jddeep** [comment]
> One way to achieve this is to pass a callback into CardDisplayWidget (a callback is a function, like onTap that's passed to GestureDetector). The callback function body will be inside CardsLearningState where it can simply call setState(), triggering build() and updating button state.
@dotdoom So can you please elaborate on how to pass the callback?

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 9 above)
Comment (dotdoom) [reply flutter/lib/views/helpers/card_display_widget.dart]: Passing callback is easy, you already do it to `GestureDetector`.

Implementing it is more involved, but here's an [example](https://github.com/flutter/flutter/blob/ab2d0323a765d297a3b7900112dbed2a562119af/examples/flutter_gallery/lib/demo/material/backdrop_demo.dart#L151). There are 2 things to do: (1) declare a `Function` type for the callback and (2) declare a field in the Widget and a parameter to constructor.

Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: @dotdoom Yeah i had done same thing earlier but it didn't work out. I guess I need to try some other way.

Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: Will tell you about it.

**Comment by jddeep** [comment]
@dotdoom Please check now, It shows an error : `setstate() or markNeedsBuild() called during build`

**Comment by jddeep** [comment]
> @dotdoom Please check now, It shows an error : `setstate() or markNeedsBuild() called during build`
@dotdoom please ignore the CI build for now and please help me in this.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 10]
```diff
@@ -93,10 +93,11 @@ class CardsLearningState extends State<CardsLearning> {
                         backgroundColor: specifyCardBackground(
                             _viewModel.deck.type, _viewModel.card.back),
                         isMarkdown: _viewModel.deck.markdown,
+                        buildbuttons: buildButtons(context),
```
Comment (dotdoom) [inline flutter/lib/views/cards_learning/cards_learning.dart]: @jddeep here's how you can add an inline comment, which will start a conversation regarding a specific line of code:

image

As the error message says, `setState` can not be called from `build`. And it's not necessary: `build` is called automatically after `setState`. So the callback should simply call `setState(() {})` without building any buttons.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 11]
```diff
@@ -133,39 +134,27 @@ class CardsLearningState extends State<CardsLearning> {
...
-          heroTag: 'turn',
-          child: const Icon(Icons.cached),
-          onPressed: () {
-            setState(() {
-              _isBackShown = true;
-            });
-          })
-    ]);
+  Function buildButtons(BuildContext context) {
+    setState(() {});
```
Comment (jddeep) [inline flutter/lib/views/cards_learning/cards_learning.dart]: @dotdoom I tried to implement what you said, Is it cool? But the error seems to persist.

Comment (jddeep) [reply flutter/lib/views/cards_learning/cards_learning.dart]: @dotdoom Can you help me with some code snippet? Thanks.

Comment (dotdoom) [reply flutter/lib/views/cards_learning/cards_learning.dart]: So there are 2 problems here:

1. `buildButtons` must return a Widget (not a Function), therefore it must end with `return SlowOperationWidget(...)` statement.

1. `setState(...)` must not be called inside `buildButtons`. It must be passed instead of `buildButtons` above:

```dart
// NOTE: no buildButtons call (just the parameter name "buildButtons", which I suggest to rename to "onFlip").
buildButtons: () => { setState((){}); }
```

Comment (jddeep) [reply flutter/lib/views/cards_learning/cards_learning.dart]: Well @dotdoom i need to use ` widget.buildButtons: ` i believe.

Comment (dotdoom) [reply flutter/lib/views/cards_learning/cards_learning.dart]: Could you please update the code, rename the `buildButtons` field to `onFlip` and let's review again?

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 12]
```diff
@@ -93,10 +93,12 @@ class CardsLearningState extends State<CardsLearning> {
                         backgroundColor: specifyCardBackground(
                             _viewModel.deck.type, _viewModel.card.back),
                         isMarkdown: _viewModel.deck.markdown,
+                        onFlip: buildButtons(context),
+                        // onFlip: ()=>{setState((){});}
```
Comment (jddeep) [inline flutter/lib/views/cards_learning/cards_learning.dart]: @dotdoom The name has been changed to `onFlip` . I have tried here what you suggested but there was an error there.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 13]
```diff
@@ -133,40 +135,24 @@ class CardsLearningState extends State<CardsLearning> {
...
-          heroTag: 'turn',
-          child: const Icon(Icons.cached),
-          onPressed: () {
-            setState(() {
-              _isBackShown = true;
-            });
-          })
-    ]);
-  }
+  Widget buildButtons(BuildContext context) => SlowOperationWidget((cb) => Row(
```
Comment (jddeep) [inline flutter/lib/views/cards_learning/cards_learning.dart]: @dotdoom This has been changed to Widget,

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 14]
```diff
@@ -1,58 +1,100 @@
...
+          builder: (context, child) => Transform(
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: () {
+                    _startAnimation();
+                    if (backshown) {
+                      // widget.onFlip;
```
Comment (jddeep) [inline flutter/lib/views/helpers/card_display_widget.dart]: @dotdoom Here it shows ` avoid using unnecessary statements` .

**Comment by jddeep** [comment]
@dotdoom After this push, I have made some new changes to the code locally like making `onFlip` a voidCallback and calling buildButtons within `setState()` . This has resulted in the application to run fine without errors but the buttons are not getting built on tap upon the card. The animation and other functionalities run fine though.

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 14 above)
Comment (dotdoom) [reply flutter/lib/views/helpers/card_display_widget.dart]: This should be `widget.onFlip()` - calling a method in Dart requires parenthesis.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 15]
```diff
@@ -93,10 +93,12 @@ class CardsLearningState extends State<CardsLearning> {
...
                         isMarkdown: _viewModel.deck.markdown,
+                        onFlip: buildButtons(context),
+                        // onFlip: ()=>{setState((){});}
                       )),
-                      Padding(
-                        padding: const EdgeInsets.only(top: 25.0, bottom: 20.0),
-                        child: _buildButtons(context),
+                      const Padding(
+                        padding: EdgeInsets.only(top: 25.0, bottom: 20.0),
+                        // child: buildButtons(context),
```
Comment (dotdoom) [inline flutter/lib/views/cards_learning/cards_learning.dart]: This must be un-commented for the buttons to build.

File: flutter/lib/views/cards_learning/cards_learning.dart (hunk 12 above)
Comment (dotdoom) [reply flutter/lib/views/cards_learning/cards_learning.dart]: You don't have to call `buildButtons`.
This statement should look like this:

```dart
onFlip: () { setState((){}); }
```

`setState()` will call `build()` method, which will update the buttons (`buildButtons`). If you remember the "counter" example (default Flutter example), there only thing that happens on FAB click is `setState()` -- no other things like `buildTextInTheMiddleOfScreen` etc.

**Comment by jddeep** [comment]
@dotdoom Please take a look now. Thanks.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 16]
```diff
@@ -93,10 +93,20 @@ class CardsLearningState extends State<CardsLearning> {
...
+                        onFlip: () => {
+                              setState(() {
+                                _isBackShown = true;
+                              })
+                            },
                       )),
                       Padding(
                         padding: const EdgeInsets.only(top: 25.0, bottom: 20.0),
-                        child: _buildButtons(context),
+                        child: _isBackShown
```
Comment (dotdoom) [inline flutter/lib/views/cards_learning/cards_learning.dart]: Please revert - this change should not be necessary.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 17]
```diff
@@ -133,40 +143,24 @@ class CardsLearningState extends State<CardsLearning> {
...
-          heroTag: 'turn',
-          child: const Icon(Icons.cached),
-          onPressed: () {
-            setState(() {
-              _isBackShown = true;
-            });
-          })
-    ]);
-  }
+  Widget _buildButtons(BuildContext context) => SlowOperationWidget((cb) => Row(
```
Comment (dotdoom) [inline flutter/lib/views/cards_learning/cards_learning.dart]: Please revert changes to `_buildButtons`.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 18]
```diff
@@ -1,58 +1,100 @@
...
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: () {
+                    _startAnimation();
+                    if (backshown) {
+                      widget.onFlip();
+                      backshown = false;
```
Comment (dotdoom) [inline flutter/lib/views/helpers/card_display_widget.dart]: Next step is to pass isBackShown as a parameter to onFlip.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 19]
```diff
@@ -1,58 +1,100 @@
...
-          ),
-        ))
-        ..add(_sideText(back, context));
-    }
+class _CardDisplayWidgetState extends State<CardDisplayWidget>
+    with SingleTickerProviderStateMixin {
+  bool flipped = false;
+  Animation<double> _animation;
+  AnimationController _controller;
+  bool backshown = false;
```
Comment (dotdoom) [inline flutter/lib/views/helpers/card_display_widget.dart]: What's the difference between `flipped` and `backshown`?

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 18 above)
Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: @dotdoom I am calling the ` onFlip ` callback only when `backshown ` . So why to pass it again as a parameter to the same?

File: flutter/lib/views/cards_learning/cards_learning.dart (hunk 17 above)
Comment (jddeep) [reply flutter/lib/views/cards_learning/cards_learning.dart]: @dotdoom It was showing an `flutter analyze` error and directed to convert into exprssion function.

Comment (dotdoom) [reply flutter/lib/views/cards_learning/cards_learning.dart]: That's because you have removed some lines from the function. Please revert.

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 18 above)
Comment (dotdoom) [reply flutter/lib/views/helpers/card_display_widget.dart]: I think `onFlip` should be called in both cases.

**Comment by jddeep** [comment]
@dotdoom All changes have been seen into. But I could not figure how to pass the `backshown` as a parameter to the `VoidCallback` `onFlip` . Please help here.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 20]
```diff
@@ -1,58 +1,94 @@
...
+          animation: _animation,
+          builder: (context, child) => Transform(
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: () {
+                    _startAnimation();
+                    widget.onFlip();
```
Comment (jddeep) [inline flutter/lib/views/helpers/card_display_widget.dart]: Like how to pass `backshown` as parameter here? @dotdoom

Comment (dotdoom) [reply flutter/lib/views/helpers/card_display_widget.dart]: You can't use VoidCallback anymore, because this predefined type doesn't carry a parameter.

You'll have to define your own Function type, OnFlipCallback. I suggest to read Dart docs on defining Function types, but here's an example:

https://github.com/dasfoo/delern/blob/efd8e038f97e80ac3c7bdd08c86e70289b2fbfc2/flutter/lib/views/deck_settings/deck_type_dropdown_widget.dart

**Comment by jddeep** [comment]
@dotdoom @ksheremet Everything works fine now. Please check it. And @dotdoom Thanks for your constant support.

**Comment by jddeep** [comment]
@ksheremet the travis CI build fails and it is showing that `flutter packages get` exits with 0. But i haven't changed the pubspec.yaml file.

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 21]
```diff
@@ -1,58 +1,96 @@
...
+          animation: _animation,
+          builder: (context, child) => Transform(
+                alignment: Alignment.center,
+                transform: Matrix4.identity()
+                  ..setEntry(3, 2, 0.001)
+                  ..rotateY(_animation.value),
+                child: GestureDetector(
+                  onTap: () {
+                    _startAnimation();
+                    widget.onFlip(backshown);
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: @jddeep It throws null pointer exception here in Card Preview:
flutter: ══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
flutter: The following NoSuchMethodError was thrown while handling a gesture:
flutter: The method 'call' was called on null.
flutter: Receiver: null
flutter: Tried calling: call(true)

Please check on null before calling the function:
```
if (widget.onFlip != null) {
   widget.onFlip(backshown);
}
```

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 22]
```diff
@@ -1,58 +1,96 @@
...
+
+  Widget _buildCard(Color backgroundcolor, String text) => Padding(
+      padding: const EdgeInsets.only(top: 30.0, bottom: 30.0),
+      child: Card(
+        color: backgroundcolor,
+        child: Container(
+          height: MediaQuery.of(context).size.height,
+          width: MediaQuery.of(context).size.width - 50.0,
+          child: ListView(scrollDirection: Axis.vertical, children: <Widget>[
+            Center(child: Text(text, style: AppStyles.primaryText))
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: @jddeep 
As I see code that supports markdown feature was deleted. Please have a look on markdown and return markdown support in cards:
```
if (isMarkdown) {
      return buildNonScrollingMarkdown(text, context);
}
```
Please have a look on all supported features in Readme

Changes that you made:
Simulator Screen Shot - iPhone XS Max - 2019-03-21 at 17 21 23

How text should look:
Simulator Screen Shot - iPhone XS Max - 2019-03-21 at 17 27 39

Please make sure that text has proper padding.

File: flutter/lib/views/cards_learning/cards_learning.dart [hunk 23]
```diff
@@ -154,18 +159,7 @@ class CardsLearningState extends State<CardsLearning> {
...
-          heroTag: 'turn',
-          child: const Icon(Icons.cached),
-          onPressed: () {
-            setState(() {
-              _isBackShown = true;
-            });
-          })
-    ]);
+    return Container();
   }
```
Comment (ksheremet) [inline flutter/lib/views/cards_learning/cards_learning.dart]: ezgif com-optimize
@jddeep 
The app has the wrong behavior. Please make sure when a user answers the card it shows always **front side of card** from the beginning.

**Comment by jddeep** [comment]
@ksheremet Please see now.

File: flutter/pubspec.yaml [hunk 24]
```diff
@@ -42,7 +42,7 @@ dev_dependencies:
   flutter_driver:
     sdk: flutter
 
-  intl_translation: ^0.17.3
+  intl_translation: ^0.17.0
```
Comment (ksheremet) [inline flutter/pubspec.yaml]: @jddeep 
Please revert these changes back in the whole file. It downgrades dependencies

**Comment by jddeep** [comment]
@ksheremet I have removed the pubspec file from commit. Please see now.

**Comment by jddeep** [comment]
@ksheremet Please take a look.

**Comment by jddeep** [comment]
@ksheremet Please review. This PR has been sitting for quite long now,

File: flutter/lib/views/cards_learning/cards_learning.dart (hunk 23 above)
Comment (ksheremet) [reply flutter/lib/views/cards_learning/cards_learning.dart]: @jddeep 

Please show only one side as it was before. Please show buttons on back side. When problem is solved please upload a gif to show that it works correctly.
ezgif com-video-to-gif (2)

**Comment by jddeep** [comment]
ezgif com-video-to-gif (1)
@ksheremet Please take a look now.

File: flutter/test/card_display_widget_test.dart [hunk 25]
```diff
@@ -24,7 +24,7 @@ void main() {
     final frontFinder = find.text(frontSide);
     final backFinder = find.text(backSide);
 
-    expect(frontFinder, findsOneWidget);
-    expect(backFinder, findsOneWidget);
+    expect(frontFinder, findsWidgets);
```
Comment (ksheremet) [inline flutter/test/card_display_widget_test.dart]: @jddeep It doesn't have a widget with front text. Please see the comments above for the explanation.

**Comment by jddeep** [comment]
@ksheremet Please review.

File: flutter/test/card_display_widget_test.dart (hunk 25 above)
Comment (jddeep) [reply flutter/test/card_display_widget_test.dart]: @ksheremet The test is still not working. What can be done?

File: flutter/test/card_display_widget_test.dart [hunk 26]
```diff
@@ -14,17 +14,26 @@ void main() {
...
+        back: backSide,
+        backgroundColor: AppStyles.cardBackgroundColors[Gender.feminine],
+        isMarkdown: false,
+        showBack: true,
+      ),
+    ));
+    final backFinder = find.widgetWithText(Widget, backSide);
+    expect(backFinder, findsWidgets);
   });
 }
```
Comment (ksheremet) [inline flutter/test/card_display_widget_test.dart]: @jddeep 
Please fix [card_preview.dart](https://github.com/dasfoo/delern/blob/438fd40bd50a9a845ae209833ab72dc42728ebb3/flutter/lib/views/card_preview/card_preview.dart#L80) 
Flipping doesn't work there anymore, because onFlip param is skipped.

File: flutter/test/card_display_widget_test.dart [hunk 27]
```diff
@@ -14,17 +14,26 @@ void main() {
...
+    await tester.pumpWidget(MaterialApp(
+      home: CardDisplayWidget(
+        back: backSide,
+        backgroundColor: AppStyles.cardBackgroundColors[Gender.feminine],
+        isMarkdown: false,
+        showBack: true,
+      ),
+    ));
+    final backFinder = find.widgetWithText(Widget, backSide);
+    expect(backFinder, findsWidgets);
```
Comment (ksheremet) [inline flutter/test/card_display_widget_test.dart]: After fixing all above the test still fails, because there is a problem in _CardDisplayWidgetState
```dart
             widget.showBack
                   ? _buildCardBody(context, widget.back)
                   : _buildCardBody(context, widget.front),
```
When card was flipped backshown was changed to false. But widget.showBack is still false.
I would recommend trying out ```backshown``` instead of ```widget.showBack```.
You can investigate the problem and see how test run in emulator by running the following command:
```
flutter run test/card_display_widget_test.dart
```

**Comment by jddeep** [comment]
@ksheremet Made changes. The test shows a new error:
``` Bad state : Too many elements ```

File: flutter/lib/views/helpers/card_display_widget.dart [hunk 28]
```diff
@@ -1,52 +1,111 @@
...
   final bool isMarkdown;
+  final onFlipCallback onFlip;
 
   const CardDisplayWidget(
-      {@required this.front,
-      @required this.back,
-      @required this.showBack,
-      @required this.backgroundColor,
-      @required this.isMarkdown});
+      {this.front,
```
Comment (ksheremet) [inline flutter/lib/views/helpers/card_display_widget.dart]: @jddeep 
Why is ```required``` gone?
Wrong formatting as well.

File: flutter/test/card_display_widget_test.dart (hunk 25 above)
Comment (ksheremet) [reply flutter/test/card_display_widget_test.dart]: @jddeep 
How about ```expect(frontFinder, findsOneWidget);```

File: flutter/test/card_display_widget_test.dart [hunk 29]
```diff
@@ -22,9 +22,12 @@ void main() {
...
-    final backFinder = find.text(backSide);
+    expect(frontFinder, findsWidgets);
+
+    await tester.tap(find.byType(AnimatedBuilder));
 
-    expect(frontFinder, findsOneWidget);
-    expect(backFinder, findsOneWidget);
+    await tester.pumpAndSettle();
+    final backFinder = find.text(backSide);
+    expect(backFinder, findsWidgets);
```
Comment (ksheremet) [inline flutter/test/card_display_widget_test.dart]: @jddeep 

Have you tried?
```dart
expect(backFinder, findsOneWidget);
```

File: flutter/test/card_display_widget_test.dart [hunk 30]
```diff
@@ -22,9 +22,12 @@ void main() {
     ));
 
     final frontFinder = find.text(frontSide);
-    final backFinder = find.text(backSide);
+    expect(frontFinder, findsWidgets);
+
+    await tester.tap(find.byType(AnimatedBuilder));
```
Comment (ksheremet) [inline flutter/test/card_display_widget_test.dart]: @jddeep 
Please try out 
```dart
find.byType(Card)
```

File: flutter/lib/views/helpers/card_display_widget.dart (hunk 28 above)
Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: @ksheremet What do you mean by wrong formatting? I had formatted the code according to dartfmt.

Comment (jddeep) [reply flutter/lib/views/helpers/card_display_widget.dart]: @ksheremet I am making all parameters required then except `onFlip` cause there is no need of it in the test.

**Comment by jddeep** [comment]
@ksheremet I made some changes. The test passes successfully now. Please review.

**Comment by jddeep** [comment]
@ksheremet All checks and builds have passed as well, Please review, :)

**Comment by ksheremet** [comment]
@jddeep
Please fix card_preview.dart
Flipping doesn't work there anymore, because onFlip param is skipped.

**Comment by ksheremet** [comment]
@jddeep 

Card Preview still doesn't work. Please fix it.

File: flutter/test/card_display_widget_test.dart [hunk 31]
```diff
@@ -17,14 +18,19 @@ void main() {
...
         showBack: true,
       ),
     ));
 
     final frontFinder = find.text(frontSide);
-    final backFinder = find.text(backSide);
-
     expect(frontFinder, findsOneWidget);
-    expect(backFinder, findsOneWidget);
+    if (backshown) {
```
Comment (ksheremet) [inline flutter/test/card_display_widget_test.dart]: @jddeep 

```backshown```  is always false. The following code will never be executed:
```dart
if (backshown) {
      await tester.tap(find.byType(Card));

      await tester.pumpAndSettle();
      final backFinder = find.text(backSide);
      expect(backFinder, findsOneWidget);
    }
```

Please debug the code, find what is wrong and fix it

**Comment by jddeep** [comment]
@ksheremet `CardPreview` works fine now. I have checked all other parts and none are broken. The test passes successfully as well. Please review.

**Comment by jddeep** [comment]
@ksheremet The ci build fails here due to `working directory not clean` . Why is that? I staged everything to commit.
